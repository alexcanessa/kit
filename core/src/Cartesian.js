import React, { Fragment } from 'react'
import { cartesianProduct } from './util'

export default ({ component, container = Fragment, ...props }) => {
  const combinations = cartesianProduct(props)
  const Component = component
  const Container = container
  
  return (
    <Fragment>
      {combinations.map((props, i) => (
        <Container key={i} >
          <Component {...props} />
        </Container>
      ))}
    </Fragment>
  )
}
