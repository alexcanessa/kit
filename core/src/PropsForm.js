import React from 'react'
import { Box, Flex, Label, Input, Select, Divider } from './ui'

class PropsForm extends React.Component {
  constructor(props) {
    super()
    const [child] = React.Children.toArray(props.children)
    const initialState = child ? child.props : {}
    this.state = initialState
    this.update = fn => this.setState(fn)
  }

  componentWillReceiveProps(next) {
    if (next.children !== this.props.children) {
      const [child] = React.Children.toArray(next.children)
      this.update(child.props)
    }
  }

  render() {
    const children = React.Children.toArray(this.props.children)
    const [child] = children
    const clone = React.cloneElement(child, this.state)

    const controls = children
      .filter(c => c.type.isForm === true)
      .map(c =>
        React.cloneElement(c, { props: this.state, update: this.update })
      )

    return (
      <Flex flexDirection="column">
        <Box pb={4} flex="1 1 auto">
          {React.cloneElement(child, this.state)}
        </Box>
        <Divider />
        <Box flex="none" py={2} mt={2}>
          {controls}
        </Box>
      </Flex>
    )
  }
}

PropsForm.Input = ({ update, name, props = {}, ...rest }) => (
  <Box mb={2}>
    <Label>{name}</Label>
    <Input
      {...rest}
      name={name}
      value={props[name] || ''}
      onChange={e => {
        const { value } = e.target
        update({ [name]: value })
      }}
    />
  </Box>
)
PropsForm.Input.isForm = true

PropsForm.Select = ({ update, name, props = {}, ...rest }) => (
  <Box mb={2}>
    <Label>{name}</Label>
    <Select
      {...rest}
      name={name}
      value={props[name] || ''}
      onChange={e => {
        const { value } = e.target
        update({ [name]: value })
      }}
    />
  </Box>
)
PropsForm.Select.isForm = true

PropsForm.Checkbox = ({ update, name, props = {}, ...rest }) => (
  <Box mb={2}>
    <Label>{name}</Label>
    <Input
      {...rest}
      name={name}
      type="checkbox"
      checked={!!props[name]}
      onChange={e => {
        const { checked } = e.target
        update({ [name]: checked })
      }}
    />
  </Box>
)
PropsForm.Checkbox.isForm = true

export default PropsForm
