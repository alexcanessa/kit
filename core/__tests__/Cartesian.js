import React from 'react';
import styled from 'styled-components';
import { render } from 'react-testing-library';
import { color, space, fontSize } from 'styled-system';

import { Cartesian } from '../src';

const buttonProps = {
  children: 'Hello, world!',
  fontSize: [1, 2, 3],
  px: [2, 3],
  backgroundColor: ['pink', 'tomato', 'purple'],
};

const Button = styled.a`
  ${color}
  ${fontSize}
  ${space}
`;

const Container = styled.div`
  padding: 1rem;
`;

test('Cartesian renders all examples', () => {
  const { container } = render(
    <Cartesian {...buttonProps} component={Button} />,
  );

  expect(container).toMatchSnapshot();
});

test('Cartesian renders all examples in a container', () => {
  const { getByTestId, container } = render(
    <Cartesian
      {...buttonProps}
      component={Button}
      container={() => <Container data-testid="container" />}
    />,
  );

  const containerNode = getByTestId('container');
  expect(containerNode.nodeName).toBe('DIV');
  expect(container).toMatchSnapshot();
});
