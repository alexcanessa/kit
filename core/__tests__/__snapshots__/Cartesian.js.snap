// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cartesian renders all examples 1`] = `
<div>
  <a
    class="sc-bdVaJa ikbxON"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa ikbxON"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa ikbxON"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa cSmzEW"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa cSmzEW"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa cSmzEW"
    font-size="1"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa hFxVwo"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa hFxVwo"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa hFxVwo"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa hPCeld"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa hPCeld"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa hPCeld"
    font-size="2"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa hJuUIp"
    font-size="3"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa hJuUIp"
    font-size="3"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa hJuUIp"
    font-size="3"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa fGNklA"
    font-size="3"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa fGNklA"
    font-size="3"
  >
    Hello, world!
  </a>
  <a
    class="sc-bdVaJa fGNklA"
    font-size="3"
  >
    Hello, world!
  </a>
</div>
`;

exports[`Cartesian renders all examples in a container 1`] = `
<div>
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
  <div
    class="sc-bwzfXH UBTUY"
    data-testid="container"
  />
</div>
`;
